_posts/2019-04-24-server-rendering-responsively.md

```diff
@@ -0,0 +1,199 @@
+app, in which case we don't have to also server-render the desktop version.
+
+TODO: is this sniffing part of react-responsive-media, or is it separate? I should find this out and make it clear.
+
+## The current state of @artsy/react-responsive-media
+
+With this, we now have:
+
+- TODO: list the features out.
+
+We're pretty happy with this, and we have generally good feelings about where it's at. We're proud of the fact that
+our approach works well for SSR, even in cases where you're doing _only_ SSR.
+
+## `styled-system`
+
+One other major factor in our approach to a responsive [artsy.net](artsy.net) is the
+[`styled-system`](https://styled-system.com/) project. We started using it around the same time we started building
+@artsy/react-responsive-media, and it has been incredibly helpful. While @artsy/react-responsive-media is great for
+rendering large sub-trees based on a matching breakpoint, `styled-system` is really great for styling individual
+components based on a matching breakpoint. For example, here's a component that would render a `div` (`Box`) with a
+width of 50% at our mobile breakpoint, and 100% for anything larger:
+
+```
+<Box width={["50%", "100%"]}>...</Box>
+```
+
+It's pretty amazing. `styled-system` allows us to use the same components at different breakpoints, but style them
+differently at each breakpoint. Without it, we were seeing _a lot_ of duplicated components via
+@artsy/react-responsive-media. With `styled-system`, we're finding that we use @artsy/react-responsive-media to
+occasionally handle significant layout differences across breakpoints, but `styled-system` more frequently for
+subtle styling differences across breakpoints.
+
+TODO: Have we already blogged about styled-system? I feel like we might have, and I should find out & link to that
+article.
+
+## Things we haven't solved
+
+As you might expect, there are still some problems we're trying to solve. Software is always a work-in-progress,
+amiright?
+
+### SEO
+
+TODO: clean up these notes
+
+- duplicate headings are likely fine (link to webmaster video)
+
+- But also, we're truncating data for the "mobile" breakpoints, and this might be affecting what gets crawled by
+  Google
+
+- [https://artsyproduct.atlassian.net/browse/PURCHASE-973](https://artsyproduct.atlassian.net/browse/PURCHASE-973)
+
+- But then, who really knows, because SEO is magic.
+
+### Accessibility
+
+TODO: clean up these notes
+
+- Modern screenreaders are also parsing semantics (html5/sections), but older screenreaders are not.
+
+- We are still making progress here, including adding the ability to specify element for typography in our design
+  system, to increase the semantic meaning of our markup.
+
+- Media emits a div, so we have to consider in sizing our components inside of a Media element
+
+## Conclusion
+
+TODO: Figure out some really smart thing to say here, or ask someone really smart to think of something really
+smart to say.
```

Comment:
Responsive design is hard, especially when layouts change significantly between desktop and mobile. Server side rendering in react is hard to get right, period. Hydration errors are easy to introduce and can break your app in weird and unexpected ways. Adding responsiveness on top of that just makes it _that_ much more complicated. 

At the end of the day, do everything you can to limit layout differences between mobile and desktop. Use `styled-systems` responsive props. Play around with flexbox's `flexDirection` and start learning about `css grid`. If you absolutely must render different views on different breakpoints, render all the UI and hide what's not needed for that breakpoint. That's the best we can do until unless the rehydration mechanism improves.